python_modules/dagster/dagster/_core/definitions/job_base.py

```diff
@@ -24,127 +19,23 @@ def get_definition(self) -> "JobDefinition":
         pass
 
     @abstractmethod
-    def subset_for_execution(
+    def get_subset(
         self,
-        solid_selection: Optional[Sequence[str]] = None,
+        op_selection: Optional[Iterable[str]] = None,
         asset_selection: Optional[AbstractSet[AssetKey]] = None,
     ) -> "IJob":
         pass
 
     @property
     @abstractmethod
-    def solids_to_execute(self) -> Optional[AbstractSet[str]]:
+    def op_selection(self) -> Optional[Sequence[str]]:
         pass
 
     @property
     @abstractmethod
     def asset_selection(self) -> Optional[AbstractSet[AssetKey]]:
         pass
 
-    @abstractmethod
-    def subset_for_execution_from_existing_job(
-        self,
-        solids_to_execute: Optional[AbstractSet[str]] = None,
-        asset_selection: Optional[AbstractSet[AssetKey]] = None,
-    ) -> "IJob":
-        pass
-
-
-class InMemoryJob(IJob, object):
```

Comment:
Having `JobDefinition` implement `IJob` directly makes it possible to pass the "unwrapped" `JobDefinition` from a  `ReconstructableJob` down the callstack and lose its reconstruction context. Worried that blurring the container type and contained type together may be more confusing.

Reply:
Currently it is possible to pass an `InMemoryJob` down the callstack with no reconstruction context-- in cases where that context is required, the type can be limited to `ReconstructableJob` rather than `IJob`.

>Worried that blurring the container type and contained type together may be more confusing.

IMO the status quo is over-complex because `JobDefinition` was already implementing subsetting methods, the seeming reason for existence of `IJob`, just with different names.

With this change there is no longer really a container/contained dichotomy-- instead `IJob` just means "some subsettable representation of a job.

@dpeng817 what do you think?

Reply:
> implementing subsetting methods, the seeming reason for existence of IJob

The reason for existence of `IJob` is in the doc block for the class 

> enables these execution APIs to operate on both in memory job
    definitions to be executed in the current process (InMemoryJob) as well as definitions that
    can be reconstructed and executed in a different process (ReconstructableJob)
    
> in cases where that context is required, the type can be limited to ReconstructableJob rather than IJob

This isn't really viable, at least with out further changes, due to how the APIs are layered/shared. Most of the execution APIs in question that operate on `IJob` are hit in both in process and multi process execution. Only once the executor definition is resolved lower in the callstack do you know whether you need reconstruction information. 
    
The subsetting APIs operating on `IJob` / `ReconstructableJob` were to be able to encode the subsetting information on the pointer object that gets passed across process boundaries. I agree this part is messy - and i believe there is inconsistency around callsites expecting to unpack an already subsetted job vs calling subset on the job definition post unpack. 


Some history on this https://dagster.phacility.com/D2739 - this current sub-optimal state was to escape an even worse previous state.

Reply:
I think `InMemoryJob` has left a bad taste in my mouth from trying to parse callsites like we find in `dagster/_utils/dagster_type.py`, where we are alternatively passing these two representations of a job to different functions, but it turns out that the `InMemoryJob` doesn't actually offer any additional functionality on top of `JobDefinition`, and on the other end of the call site there is just an `isinstance` check for `ReconstructableJob` and then the rest of the logic is accessing the underlying `JobDefinition` of the passed `InMemoryJob` (this is what happens inside `create_execution_plan`).

```
    job = InMemoryJob(GraphDefinition(node_defs=[], name="empty").to_job())
    job_def = job.get_definition()

    instance = DagsterInstance.ephemeral()
    execution_plan = create_execution_plan(job)
    dagster_run = instance.create_run_for_job(job_def)
    with scoped_job_context(execution_plan, job, {}, dagster_run, instance) as context:
```

It's possible I haven't fully grasped things, but having spent a lot of time looking at these codepaths recently, I have a strong opinion that "net confusion" is reduced by simplifying sites like the above by eliminating `InMemoryJob`. I think it would be different if `InMemoryJob` was used more like an "over-the-wire" representation, where we wrapped a `JobDefinition` in it, passed it through some code that only accessed limited `IJob` APIs, and then "unwrapped it" into a `JobDefinition` at the end. But instead there's a lot of intermittent access to the definition, which makes `InMemoryJob` feel like an unnecessary wrapper-- an abstraction barrier that isn't really respected.

That said I think you know these APIs better than I do-- if you'd prefer to keep `InMemoryJob` I can revert that part of this PR and just do the subset stuff.

Reply:
I think reverting it out of this PR is right since the rest of the changes are not contended and good.

Reply:
I agree with your assessment that `InMemoryJob` sucks. Its only reason for existence is to make it harder to mess up with `ReconstructableJob` in the core execution APIs. I am sure there is a better setup, but I'm not yet sold on this proposed approach.